### ISSUE TITLE
Unable to seed giant torrents: ProtocolException: Invalid message length received

### ISSUE THREAD

**Issue by lostmsu** [alanmcgovern/monotorrent#653]
MonoTorrent drops all incoming connections with

> MonoTorrent.Client.ProtocolException: Invalid message length received. Value was 'XXXXXX'

I suspect this has to do with the fact, that a torrent like this has bitfield that is larger than `PeerIO.MaxMessageLength`.

Example torrent: https://1drv.ms/u/s!AioR7aFFo_FK1LdTfRnaXd70hiYopw?e=WV7OS6

**Comment by alanmcgovern** [comment]
That is definitely larger than any torrent I've ever seen.

QBIttorrent won't open it:
<img width="370" alt="too_large" src="https://github.com/alanmcgovern/monotorrent/assets/264396/491d2738-7f80-426c-8f99-e5823b1daa15">

uTorrent throws an error:
<img width="257" alt="utorrent_too_large" src="https://github.com/alanmcgovern/monotorrent/assets/264396/1b7c90cc-02c6-4903-b0e6-cde9d645b5ad">

The torrent has 3,848,104 pieces, which means each BitField message is about 481kB, which is far larger than the hardcoded limit of 64kB.

There are two things here - firstly I do think the hardcoded limit needs to be raised so that all loadable torrents are able to transfer the BitField message without the connection being closed.

Separately, I do think there should be a reasonable limit on the torrent size. I suspect the 'best' approach would be to have MonoTorrent also raise an error on similarly improper torrents.

On a related note - this is 60 gigabytes of data converted into a torrent which uses 16kB pieces. This means every bittorrent piece is exactly 1 block in size. Large torrents like this typically use much larger piece sizes so the metadata is much much smaller. For 60G of data, you should really be using 2MB or 4MB pieces, not 16kB pieces.

Did you create this torrent to demonstrate a point, or is this a legit real-world torrent that is actively in use? If so, what client can load it and successfully seed/leech it?

**Comment by lostmsu** [comment]
I am using these torrents internally with a fork of MonoTorrent for client.

There are multiple reasons to set block size == piece size in my scenario (virtual file system):
1. I replace the block storage with my own implementation. It gets complicated very quickly if it can't verify the piece as soon as it is received because I need to store incomplete blocks somewhere until I can verify them. On big torrents that can bloat memory with useless pieces that do not add up to whole blocks. Keeping them in limited RAM is a DoS vector.
2. The backing storage is essentially `{ hash: data }` cache, and it is much easier to work with when all `data` blocks have the same size.

**Comment by lostmsu** [comment]
BTW, in my fork I made a simple fix where the message size limit depends on the number of blocks in the torrent. See 9c3cb6b2470bc4e247cf0961024682b125e4c3f5

TL;DR; `int maxLength = Math.Max(MaxMessageLength, pieceCount / 4);`

**Comment by alanmcgovern** [comment]
> I replace the block storage with my own implementation. It gets complicated very quickly if it can't verify the piece as soon as it is received because I need to store incomplete blocks somewhere until I can verify them. On big torrents that can bloat memory with useless pieces that do not add up to whole blocks. Keeping them in limited RAM is a DoS vector

MonoTorrent does a few things to make this not really an issue:
1) By default all blocks from a single piece are requested from the same peer. This increases the likelihood that all blocks will arrive in-order.
2) Incremental Hashing, which means the piece hash can be computed as each block arrives. Once a block is incrementally hashed, it's flushed to disk.
3) A memory cache used to buffer blocks if they arrive out of order, to reduce the frequency of writing a piece to disk, and then reading it back later.

> I replace the block storage with my own implementation... because I need to store incomplete blocks somewhere until I can verify them

Why don't you do what MonoTorrent does - store the incomplete blocks in their final destination if they have to be flushed to disk before they're validated. If the data ends up being corrupt, all blocks will be re-requested and the new ones will be written to disk (replacing the corrupt data).

Is it an issue if your flush incomplete blocks to disk  before they've been validated? If so - why? If it is a problem, can you rely on `TorrentManager.Files[index].BitField` to know whether or a given range of bytes has been validated?

**Comment by lostmsu** [comment]
> Why don't you do what MonoTorrent does - store the incomplete blocks in their final destination

This is due to n2 above: the cache contract requires contents under `hash` to have hash == `hash`. If I store incomplete block, it violates the contract.

> If it is a problem, can you rely on TorrentManager.Files[index].BitField to know whether or a given range of bytes has been validated?

The cache does not and can not know anything about torrents. It may be, that when I add a new torrent the cache already has some of its blocks either from other torrents or from past life of some kind.

> 100MB byte array in memory for the hashes

This is nothing in comparison with content I am serving, so not really a concern.

> piece picking is godawfully slow

I have a custom piece picker

**Comment by alanmcgovern** [comment]
This should work in master now! Do let me know how you get on!
